### PR TITLE
Enum assertions

### PR DESCRIPTION
**Pull request by orklah**
This will fix https://github.com/vimeo/psalm/issues/7654

It doesn't adress the collapsing issue, but it makes sure the assertions are correctly reconciled

### CODE REVIEW

File: src/Psalm/Internal/Provider/StatementsProvider.php [hunk 1]
```diff
@@ -155,10 +155,6 @@ public function getStatementsForFile(
 
             $existing_statements = $this->parser_cache_provider->loadExistingStatementsFromCache($file_path);
 
-            if ($existing_statements && !$existing_statements[0] instanceof PhpParser\Node\Stmt) {
```
Comment (orklah) [inline src/Psalm/Internal/Provider/StatementsProvider.php:158]: I made a check that went a little too far so I had to revert it, but it saw this as redundant. I couldn't find the reason for this, it seems redundant indeed

File: tests/EnumTest.php [hunk 2]
```diff
@@ -405,6 +405,35 @@ function foo(): int|Code|null
...
+
+                    $code = foo();
+                    $code1 = null;
+                    $code2 = null;
+                    if($code instanceof Code){
+                        $code1 = $code;
+                    }
+                    if(!$code instanceof Code){
+                        $code2 = $code;
+                    }',
```
Comment (weirdan) [inline tests/EnumTest.php:429]: Does it actually test reconciliation with `Enum|Enum::Case`? It doesn't seem like an individual enum case would ever appear in this test case.

Comment (orklah) [reply tests/EnumTest.php:429]: what I wanted to test was the enum case not leaking to the negated if like in here: https://github.com/vimeo/psalm/issues/7654#issuecomment-1037378155

The test would be cleaner without null values, (it showed Code|Code::Ok and never on the other var before) but assertions doesn't work if the var appears only in an if

Comment (weirdan) [reply tests/EnumTest.php:429]: I mean, this test passes on 5.0a1 (I just tested that). So it doesn't cover changes introduced in this PR.

Comment (orklah) [reply tests/EnumTest.php:429]: Ugh :( I'll try to design a better test then!

Comment (weirdan) [reply tests/EnumTest.php:429]: What you could do instead is to abort (exit or throw) in the `if` body, and then assert the type as it was changed below the `if`. E.g.
```php
[
            'EnumCaseReconciliation' => [
                'code' => '<?php
                    enum Code: int
                    {
                        case Ok = 0;
                        case Fatal = 1;
                    }

                    function foo(): Code|null|int
                    {
                        return null;
                    }

                    $code = foo() ?? Code::Ok;
                    if (!$code instanceof Code) throw new RuntimeException;
                    ',
                'assertions' => [
                    '$code===' => 'Code',
                ],
                'ignored_issues' => [],
                'php_version' => '8.1',
            ],
]
```

**Pull request merged**